### PR TITLE
[browser] Minimal blazor.boot.json integration

### PR DESCRIPTION
**Pull request by maraf**
- Support side-by-side loading from mono-config.json and blazor.boot.json produced by Wasm SDK
- This is a first step in JS/TS integration with blazor booting
- ~Updates wasmbrowser template to use Wasm SDK~ (next cycle, needs wasmapphost+wbt updates)
- Integrates `WebAssemblyResourceLoader` from Blazor
- Missing blazor features that will be considered in follow-ups
  - appSettings, reason: we don't have a C# API to actualy use them
  - JSInitializers, reason: too complicated at the moment

### CODE REVIEW

File: src/mono/wasm/runtime/blazor/_Integration.ts [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        environmentVariables["LANG"] = `${resourceLoader.startOptions.applicationCulture}.UTF-8`;
+    }
+
+    if (resourceLoader.bootConfig.aspnetCoreBrowserTools) {
+        // See https://github.com/dotnet/aspnetcore/issues/37357#issuecomment-941237000
+        environmentVariables["__ASPNETCORE_BROWSER_TOOLS"] = resourceLoader.bootConfig.aspnetCoreBrowserTools;
+    }
+}
+
+function getICUResourceName(bootConfig: BootJsonData, culture: string | undefined): string {
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/blazor/_Integration.ts:165]: We probably already have this code in icu.ts

File: src/mono/wasm/runtime/blazor/WebAssemblyConfigLoader.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            .filter(name => name === "appsettings.json" || name === `appsettings.${bootConfigResult.applicationEnvironment}.json`)
+            .map(async name => ({ name, content: await getConfigBytes(name) })));
+
+        const anyConfig = Module.config as any;
+        if (anyConfig.onAppConfigLoaded) {
+            anyConfig.onAppConfigLoaded(configFiles);
+        }
+
+        // TODO MF: Hook getConfig
+        // Blazor._internal.getConfig = (fileName: string) : Uint8Array | undefined => {
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/blazor/WebAssemblyConfigLoader.ts:23]: I think we want to include this logic as Icall in the `System.Runtime.InteropServices.JavaScript` not as hook back to blazor.

**Review by pavelsavara — COMMENTED** [review]
I would prefer to dissolve this new code into our folder structure and code style and drop `blazor` folder.

File: src/mono/wasm/runtime/run-outer.ts [hunk 3]
```diff
@@ -281,6 +282,12 @@ class HostBuilder implements DotnetHostBuilder {
         }
     }
 
+    withStartupOptions(startupOptions: Partial<WebAssemblyStartOptions>): DotnetHostBuilder {
+        const configInternal = this.moduleConfig.config as MonoConfigInternal;
+        configInternal.startupOptions = startupOptions;
+        return this.withConfigSrc("blazor.boot.json");
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/mono/wasm/runtime/run-outer.ts:288]: I prefer leave it as is at moment, close the full circle and that do changes. I believe we will end with `boot.json` for both vanilla wasm and blazor

**Comment by maraf** [comment]
> I would prefer to dissolve this new code into our folder structure and code style and drop `blazor` folder.

I definitely agree. The consider current state a WIP

File: src/mono/wasm/runtime/blazor/_Polyfill.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import { BootJsonData } from "./BootConfig";
+
+let testAnchor: HTMLAnchorElement;
+export function toAbsoluteUri(relativeUri: string): string {
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/mono/wasm/runtime/blazor/_Polyfill.ts:4]: Good catch, I haven't tried running it on console yet

File: src/mono/wasm/runtime/blazor/WebAssemblyConfigLoader.ts (hunk 2 above)
Comment (maraf) [reply src/mono/wasm/runtime/blazor/WebAssemblyConfigLoader.ts:23]: Do you think that it is possible to do it separately? Or is it that simple that we should do it right away?

File: src/mono/wasm/runtime/blazor/WebAssemblyResourceLoader.ts [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    if (window.isSecureContext === false) {
+        return null;
+    }
+
+    // Define a separate cache for each base href, so we're isolated from any other
+    // Blazor application running on the same origin. We need this so that we're free
+    // to purge from the cache anything we're not using and don't let it keep growing,
+    // since we don't want to be worst offenders for space usage.
+    const relativeBaseHref = document.baseURI.substring(document.location.origin.length);
+    const cacheName = `blazor-resources-${relativeBaseHref}`;
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/mono/wasm/runtime/blazor/WebAssemblyResourceLoader.ts:190]: I believe having `dotnet` prefix is good enough

File: src/mono/wasm/runtime/blazor/_Polyfill.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (maraf) [reply src/mono/wasm/runtime/blazor/_Polyfill.ts:4]: I keen to not have this experience (=Wasm SDK + boot.json) on node at the moment

File: src/mono/wasm/runtime/blazor/_Integration.ts [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+        }
+    }
+    const applicationCulture = resourceLoader.startOptions.applicationCulture || (navigator.languages && navigator.languages[0]);
+    const icuDataResourceName = getICUResourceName(resourceLoader.bootConfig, applicationCulture);
+    let hasIcuData = false;
+    for (const name in resources.runtime) {
+        const behavior = behaviorByName(name) as any;
+        if (behavior === "icu") {
+            if (resourceLoader.bootConfig.icuDataMode === ICUDataMode.Invariant) {
+                continue;
```
Comment (ilonatommy) [inline src/mono/wasm/runtime/blazor/_Integration.ts:137]: If mode is invariant we wouldn't like to `assets.push(assset)`, finish early:
```suggestion
                break;
```
Or even better: skip entering the loop if `(resourceLoader.bootConfig.icuDataMode === ICUDataMode.Invariant)`.

**Review by pavelsavara — APPROVED** [review]
LGTM as work in progress, expecting further cleanup and unification in subsequent PRs

File: src/mono/wasm/runtime/blazor/_Integration.ts (hunk 6 above)
Comment (maraf) [reply src/mono/wasm/runtime/blazor/_Integration.ts:137]: The loop is over all resources, it can contain also `js-module-threads`

Comment (ilonatommy) [reply src/mono/wasm/runtime/blazor/_Integration.ts:137]: Never mind, read it wrong.

**Comment by maraf** [comment]
Failures are not related

**Pull request merged**